File: articles/prenatal-development-the-earliest-foundations-of-life/python/model_prenatal_development.py
#!/usr/bin/env python3
"""Model synthetic prenatal development conditions and early outcomes."""

from __future__ import annotations

from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "prenatal_development_foundations_panel.csv"
OUTPUTS_DIR = ROOT / "outputs"
OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)


def main() -> None:
    if not DATA_PATH.exists():
        raise FileNotFoundError("Run python/generate_prenatal_development_panel.py first.")

    prenatal = pd.read_csv(DATA_PATH)

    model = smf.ols(
        """
        early_outcome ~ gestational_weeks + maternal_health + effective_care +
        nutrition_support + social_support + chronic_stress + toxic_exposure +
        environmental_burden + economic_security +
        maternal_health:effective_care +
        maternal_health:chronic_stress +
        developmental_risk:effective_care
        """,
        data=prenatal,
    ).fit(cov_type="HC3")

    neighborhood_model = smf.ols(
        """
        early_outcome ~ gestational_weeks + maternal_health + prenatal_care +
        healthcare_access + nutrition_support + social_support +
        developmental_risk + effective_care + C(prenatal_profile)
        """,
        data=prenatal,
    ).fit(cov_type="HC3")

    with open(OUTPUTS_DIR / "python_prenatal_development_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("PRENATAL DEVELOPMENT EARLY OUTCOME MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(model.summary().as_text())
        f.write("\n\nNEIGHBORHOOD AND PROFILE MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(neighborhood_model.summary().as_text())

    prenatal["stress_group"] = pd.qcut(
        prenatal["chronic_stress"],
        4,
        labels=["Lowest stress", "Moderate-low stress", "Moderate-high stress", "Highest stress"],
    )

    stress_summary = prenatal.groupby("stress_group", as_index=False, observed=True).agg(
        average_outcome=("early_outcome", "mean"),
        average_care=("effective_care", "mean"),
        average_risk=("developmental_risk", "mean"),
        standard_error=("early_outcome", lambda x: x.std() / (len(x) ** 0.5)),
    )
    stress_summary["lower"] = stress_summary["average_outcome"] - 1.96 * stress_summary["standard_error"]
    stress_summary["upper"] = stress_summary["average_outcome"] + 1.96 * stress_summary["standard_error"]
    stress_summary.to_csv(OUTPUTS_DIR / "python_prenatal_stress_summary.csv", index=False)

    plt.figure(figsize=(8, 5))
    x = stress_summary["stress_group"].astype(str)
    plt.plot(x, stress_summary["average_outcome"], marker="o")
    plt.fill_between(x, stress_summary["lower"], stress_summary["upper"], alpha=0.15)
    plt.xlabel("Prenatal stress group")
    plt.ylabel("Average early developmental outcome")
    plt.title("Synthetic Prenatal Stress and Early Developmental Outcome")
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_prenatal_stress_summary.png", dpi=160)
    plt.close()

    prenatal["care_group"] = pd.qcut(
        prenatal["effective_care"],
        4,
        labels=["Lowest care", "Moderate-low care", "Moderate-high care", "Highest care"],
    )

    care_summary = prenatal.groupby("care_group", as_index=False, observed=True).agg(
        average_outcome=("early_outcome", "mean"),
        average_risk=("developmental_risk", "mean"),
        average_gestation=("gestational_weeks", "mean"),
        average_maternal_health=("maternal_health", "mean"),
    )
    care_summary.to_csv(OUTPUTS_DIR / "python_effective_care_summary.csv", index=False)

    plt.figure(figsize=(8, 5))
    plt.plot(care_summary["care_group"].astype(str), care_summary["average_outcome"], marker="o")
    plt.xlabel("Effective prenatal care group")
    plt.ylabel("Average early developmental outcome")
    plt.title("Synthetic Effective Prenatal Care and Early Outcome")
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_effective_care_summary.png", dpi=160)
    plt.close()

    profile_summary = prenatal.groupby("prenatal_profile", as_index=False).agg(
        cases=("case_id", "count"),
        average_outcome=("early_outcome", "mean"),
        average_effective_care=("effective_care", "mean"),
        average_developmental_risk=("developmental_risk", "mean"),
        average_gestation=("gestational_weeks", "mean"),
        average_maternal_health=("maternal_health", "mean"),
        average_nutrition=("nutrition_support", "mean"),
        average_social_support=("social_support", "mean"),
    )
    profile_summary.to_csv(OUTPUTS_DIR / "python_prenatal_profiles.csv", index=False)

    print("Wrote Python prenatal development model outputs.")


if __name__ == "__main__":
    main()

File: articles/prenatal-development-the-earliest-foundations-of-life/r/model_prenatal_development.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)

data_path <- file.path(root, "data", "prenatal_development_foundations_panel.csv")
outputs_dir <- file.path(root, "outputs")

if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)
if (!file.exists(data_path)) stop("Run python/generate_prenatal_development_panel.py first.")

prenatal <- read.csv(data_path)

model <- lmer(
  early_outcome ~ gestational_weeks + maternal_health + effective_care +
    nutrition_support + social_support + chronic_stress + toxic_exposure +
    environmental_burden + economic_security +
    maternal_health:effective_care +
    maternal_health:chronic_stress +
    developmental_risk:effective_care +
    (1 | neighborhood_context),
  data = prenatal
)

capture.output(
  summary(model),
  file = file.path(outputs_dir, "r_prenatal_development_model_summary.txt")
)

stress_summary <- prenatal |>
  mutate(stress_group = ntile(chronic_stress, 4)) |>
  group_by(stress_group) |>
  summarize(
    average_outcome = mean(early_outcome),
    average_care = mean(effective_care),
    average_risk = mean(developmental_risk),
    standard_error = sd(early_outcome) / sqrt(n()),
    lower = average_outcome - 1.96 * standard_error,
    upper = average_outcome + 1.96 * standard_error,
    .groups = "drop"
  ) |>
  mutate(group = case_when(
    stress_group == 1 ~ "Lowest stress",
    stress_group == 2 ~ "Moderate-low stress",
    stress_group == 3 ~ "Moderate-high stress",
    TRUE ~ "Highest stress"
  ))

write.csv(stress_summary, file.path(outputs_dir, "r_prenatal_stress_summary.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_prenatal_stress_summary.png"),
  plot = ggplot(stress_summary, aes(x = group, y = average_outcome, group = 1)) +
    geom_line(linewidth = 1) +
    geom_point(size = 2) +
    geom_errorbar(aes(ymin = lower, ymax = upper), width = 0.15) +
    labs(title = "Synthetic Prenatal Stress and Early Developmental Outcome", x = "Prenatal stress group", y = "Average early outcome") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

care_summary <- prenatal |>
  mutate(care_group = ntile(effective_care, 4)) |>
  group_by(care_group) |>
  summarize(
    average_outcome = mean(early_outcome),
    average_risk = mean(developmental_risk),
    average_gestation = mean(gestational_weeks),
    average_maternal_health = mean(maternal_health),
    .groups = "drop"
  )

write.csv(care_summary, file.path(outputs_dir, "r_effective_care_summary.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_effective_care_summary.png"),
  plot = ggplot(care_summary, aes(x = care_group, y = average_outcome)) +
    geom_line(linewidth = 1) +
    geom_point(size = 2) +
    labs(title = "Synthetic Effective Prenatal Care and Early Outcome", x = "Effective care quartile", y = "Average early outcome") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R prenatal development outputs.")
